cpp/arcticdb/version/version_map_batch_methods.hpp

```diff
@@ -173,29 +173,89 @@ struct MapRandomAccessWrapper {
 inline std::shared_ptr<std::unordered_map<StreamId, AtomKey>> batch_get_specific_version(
     const std::shared_ptr<Store>& store,
     const std::shared_ptr<VersionMap>& version_map,
     const std::map<StreamId, VersionId>& sym_versions,
-    bool include_deleted = true) {
+    bool include_deleted = true,
+    bool include_referenced_in_snapshot_only = false) {
     ARCTICDB_SAMPLE(BatchGetLatestVersion, 0)
     auto output = std::make_shared<std::unordered_map<StreamId, AtomKey>>();
-    auto mutex = std::make_shared<std::mutex>();
+    auto output_mutex = std::make_shared<std::mutex>();
+    auto tombstoned_vers = std::make_shared<std::vector<std::pair<StreamId, AtomKey>>>(); // symbol, version key
+    auto tombstoned_vers_mutex = std::make_shared<std::mutex>();
 
     MapRandomAccessWrapper wrapper{sym_versions};
     submit_tasks_for_range(wrapper, [store, version_map](auto& sym_version) {
             LoadParameter load_param{LoadType::LOAD_DOWNTO, static_cast<SignedVersionId>(sym_version.second)};
             return async::submit_io_task(CheckReloadTask{store, version_map, sym_version.first, load_param});
         },
-        [output, include_deleted, mutex](auto sym_version, const std::shared_ptr<VersionMapEntry>& entry) {
-        auto index_key = find_index_key_for_version_id(sym_version.second, entry, include_deleted);
-        if (index_key) {
-            std::lock_guard lock{*mutex};
-            (*output)[sym_version.first] = *index_key;
+        [output, include_deleted, include_referenced_in_snapshot_only, output_mutex, store, tombstoned_vers, tombstoned_vers_mutex](auto sym_version, const std::shared_ptr<VersionMapEntry>& entry) {
+        auto key_and_is_tombstoned = find_index_key_for_version_id_and_tombstone_status(sym_version.second, entry);
+        if (key_and_is_tombstoned) {
+            bool is_valid_key = include_deleted || !key_and_is_tombstoned->second;
+            if (!is_valid_key && include_referenced_in_snapshot_only && key_and_is_tombstoned->second) { // Need to allow tombstoned version but referenced in other snapshot(s) can be "re-snapshot"
+                log::version().warn("Version {} for symbol {} is tombstoned, need to check snapshots (this can be slow)", sym_version.second, sym_version.first);
+                std::lock_guard lock{*tombstoned_vers_mutex};
+                tombstoned_vers->emplace_back(sym_version.first, key_and_is_tombstoned->first);
+            }
+            if (is_valid_key) {
+                std::lock_guard lock{*output_mutex};
+                (*output)[sym_version.first] = key_and_is_tombstoned->first;
+            }
         }
     });
 
+    if (!tombstoned_vers->empty()) {
+        auto snap_map = get_master_snapshots_map(store);
+        submit_cpu_task_for_range(*tombstoned_vers, [&snap_map = std::as_const(snap_map), &sym_versions = std::as_const(sym_versions), output, output_mutex](const auto& tombstoned_ver) {
```

Comment:
I am not sure do you mean? 
Are you saying you prefer me putting the logic in `submit_cpu_task_for_range` back to the caller? Now it is already a bunch of futures being collected at the end

Reply:
I guess I'm saying I don't understand why we have the submit_cpu_task_for_range indirection at all rather than just using the Folly APIs immediately?

Reply:
Because the logic of handling Folly APIs is irrelevant to the tombstone handling logic. So I think it is better to put the Folly APIs logic elsewhere to make it neater?
Though I agree it is a bit over-designed on `IterationTask`. It will be neater if the result is collected at the end.